mentat/conversation.py

```diff
@@ -269,3 +269,76 @@ async def get_model_response(self) -> list[FileEdit]:
         )
         self.add_model_message(parsed_llm_response.full_response, messages_snapshot)
         return parsed_llm_response.file_edits
+
+    def remaining_context(self) -> int | None:
+        ctx = SESSION_CONTEXT.get()
+        max_context = model_context_size(ctx.config.model)
+        if max_context is None:
+            return None
+
+        return max_context - conversation_tokens(self.get_messages(), ctx.config.model)
+
+    def can_add_to_context(self, message: str) -> bool:
+        """
+        Whether or not the model has enough context remaining to add this message.
+        Will take token buffer into account and uses full_message=True.
+        """
+        ctx = SESSION_CONTEXT.get()
+
+        remaining_context = self.remaining_context()
+        return (
+            remaining_context is not None
+            and remaining_context
+            - count_tokens(message, ctx.config.model, full_message=True)
+            - ctx.config.token_buffer
+            > 0
+        )
+
+    async def run_command(self, command: list[str]) -> bool:
```

Comment:
In my opinion this logic should be in the `RunCommand` and it should call `add_system_message` (which doesn't exist but can be added as part of this pull request.

Reply:
Hmm, well we're going to want this logic in the future for other things (like agent mode), so I don't think it makes sense to be command specific. The reason we don't have an add_system_message is because the add_model_message and add_user_message are specifically for messages sent by the user/model so that the transcript viewer (on the left) will look like the user expects it to; we could add an add_system_message function, but I think it would obscure the purpose of the model and user message functions (sometimes we still add model and user messages with the add_message function if the user or model didn't actually send them, so we don't want them to appear on the left of the transcript viewer)

Reply:
That makes sense to me. Thanks.